feat(s3-deployment): allow ability to specify versionId for S3 bucket source

nritholtz opened the pull request:
### Issue # (if applicable)

Closes #32462.

### Reason for this change

Allow user to specify optional versionId of the zipped file for the s3 bucket source type.

### Description of changes

1. Added optional `versionId` to [S3 Bucket source properties](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3_deployment.SourceConfig.html) and new `BucketSourceOptions` interface. 
2. Changed the underlying CLI commands to use `s3api get-object` instead of  `aws cp` to allow support of specifying versionId when supplied.

### Description of how you validated changes

Added and updated appropriate unit/integration tests. Also needed to update `solutionStackName: '64bit Amazon Linux 2023 v6.2.2 running Node.js 20',` since no longer available.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.
	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 's3-deployment' instead of 'aws-s3-deployment'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation reviewed (dismissed):
The pull request linter has failed. See the aws-cdk-automation comment below for failure reasons. If you believe this pull request should receive an exemption, please comment and provide a justification.


A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed add `Clarification Request` to a comment.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.
	❌ Features must contain a change to an integration test file and the resulting snapshot.
	❌ The title of the pull request should omit 'aws-' from the name of modified packages. Use 's3-deployment' instead of 'aws-s3-deployment'.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to a README file.
	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 72523fe38d540564c84110cccfe7a46dd9acbec8
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2d6cf197-4331-48a6-ad24-6d40ee7245f9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b523065a1012d6afacb6fd643915cf66dda7501e
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a0b905d6-2760-4d3c-9274-f075eaf33af9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 7e93569821d2d56bd9632ce49c5587926b233c10
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e156abaa-2f3f-47c0-921b-20cc64fbec28%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

nritholtz commented:
First of all, if someone can please rerun/regenerate the integration snapshots for:
```
yarn integ test/aws-codepipeline-actions/test/integ.pipeline-elastic-beanstalk-deploy.js test/aws-dynamodb/test/integ.import-source.js test/aws-ecs/test/ec2/integ.environment-file.js test/aws-s3-deployment/test/integ.bucket-deployment.js test/aws-servicecatalog/test/integ.product.js test/aws-servicecatalog/test/integ.two-products.js --update-on-failed
```
For various org reasons, I cannot run these integ tests. They have the same diff for the new custom_resource_function key signature:
```
Verifying integration test snapshots...

  CHANGED    aws-servicecatalog/test/integ.product 7.022s
      Resources
[~] AWS::Lambda::Function CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536
 └─ [~] Code
     └─ [~] .S3Key:
         ├─ [-] c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip
         └─ [+] 44d0acd3c337834b0a8b4fe963cfd68226141ad00b4a43fa3ba643546cb7576d.zip


  CHANGED    aws-codepipeline-actions/test/integ.pipeline-elastic-beanstalk-deploy 7.305s
      Resources
[~] AWS::Lambda::Function CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536
 └─ [~] Code
     └─ [~] .S3Key:
         ├─ [-] c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip
         └─ [+] 44d0acd3c337834b0a8b4fe963cfd68226141ad00b4a43fa3ba643546cb7576d.zip


  CHANGED    aws-dynamodb/test/integ.import-source 7.216s
      Resources
[~] AWS::Lambda::Function CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536
 └─ [~] Code
     └─ [~] .S3Key:
         ├─ [-] c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip
         └─ [+] 44d0acd3c337834b0a8b4fe963cfd68226141ad00b4a43fa3ba643546cb7576d.zip


  CHANGED    aws-servicecatalog/test/integ.two-products 7.32s
      Resources
[~] AWS::Lambda::Function CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C256MiB5F7BD223
 └─ [~] Code
     └─ [~] .S3Key:
         ├─ [-] c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip
         └─ [+] 44d0acd3c337834b0a8b4fe963cfd68226141ad00b4a43fa3ba643546cb7576d.zip


  CHANGED    aws-ecs/test/ec2/integ.environment-file 7.649s
      Resources
[~] AWS::Lambda::Function CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536
 └─ [~] Code
     └─ [~] .S3Key:
         ├─ [-] c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip
         └─ [+] 44d0acd3c337834b0a8b4fe963cfd68226141ad00b4a43fa3ba643546cb7576d.zip


  CHANGED    aws-s3-deployment/test/integ.bucket-deployment 7.591s
      Resources
[~] AWS::Lambda::Function CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536
 └─ [~] Code
     └─ [~] .S3Key:
         ├─ [-] c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip
         └─ [+] 44d0acd3c337834b0a8b4fe963cfd68226141ad00b4a43fa3ba643546cb7576d.zip
[~] AWS::Lambda::Function CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756Cc84ffe8a457bee0fd39689d96b787dc2cec93d55901F85F433
 └─ [~] Code
     └─ [~] .S3Key:
         ├─ [-] c6358465bf49dfae556bb430bf9c81fa578c221b82c308e3707901b1dd654762.zip
         └─ [+] 44d0acd3c337834b0a8b4fe963cfd68226141ad00b4a43fa3ba643546cb7576d.zip
```

Exemption Request for adding new integration test. It would require having an existing object in an s3 bucket, which I'm not sure how to do here. I looked in the package to see any existing integration test coverage for other source.Bucket types and I don't see any.

aws-cdk-automation commented:
The pull request linter fails with the following errors:

	❌ Features must contain a change to an integration test file and the resulting snapshot.

<b>PRs must pass status checks before we can provide a meaningful review.</b>

If you would like to request an exemption from the status checks or clarification on feedback, please leave a comment on this PR containing `Exemption Request` and/or `Clarification Request`.

✅ A exemption request has been requested. Please wait for a maintainer's review.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 1905ed8626287db9a39b93fa6a5e4c97db027479
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3547e176-0d14-4553-ac1e-3b2a3e3e9376%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4c0e0fb6dced619ec5b73893cd8c5dd60d8d6dd6
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=227b549e-b7c5-4dd1-813d-2dab8d258dc4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
<!--STALE PR-->
This PR has been in the MERGE CONFLICTS state for 3 weeks, and looks abandoned. To keep this PR from being closed, please continue work on it. If not, it will automatically be closed in a week.

aws-cdk-automation commented:
This PR has been deemed to be abandoned, and will be automatically closed. Please create a new PR for these changes if you think this decision has been made in error.

aaythapa commented:
Hi @nritholtz thank you for opening this PR! Sorry about the late response. Looks like it got close because it was deemed that it was abandoned because there was some merge conflicts. Not that if the BUILD is failing or if there are merge conflicts then the PR won't show up on our board to review. I've re-opened the PR and if you're still working on it and can fix the conflicts then we'd be happy to take a look after. Thanks again for your contribution.

aws-cdk-automation commented:
This PR has been deemed to be abandoned, and will be automatically closed. Please create a new PR for these changes if you think this decision has been made in error.
